pkg/server/modules/modules.go

```diff
@@ -0,0 +1,163 @@
+package modules
+
+import (
+	"context"
+
+	"github.com/grafana/dskit/modules"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/server/backgroundsvcs"
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/grpcserver"
+	"github.com/grafana/grafana/pkg/services/store/kind"
+	objectdummyserver "github.com/grafana/grafana/pkg/services/store/object/dummy"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const (
+	All                   string = "all"
+	GRPCServer            string = "grpc-server"
+	GRPCServerHealthCheck string = "grpc-server-health-check"
+	GRPCServerReflection  string = "grpc-server-reflection"
+	Core                  string = "core"
+	ObjectStore           string = "object-store"
+)
+
+type Modules struct {
+	cfg *setting.Cfg
+	log log.Logger
+
+	grpcServer                grpcserver.Provider
+	kindRegistry              kind.KindRegistry
+	backgroundServiceRegistry *backgroundsvcs.BackgroundServiceRegistry
+
+	ModuleManager  *modules.Manager
+	ServiceManager *services.Manager
+	serviceMap     map[string]services.Service
+	deps           map[string][]string
+}
+
+func ProvideService(cfg *setting.Cfg, server grpcserver.Provider, kindRegistry kind.KindRegistry, backgroundServiceRegistry *backgroundsvcs.BackgroundServiceRegistry, roleRegistry accesscontrol.RoleRegistry) *Modules {
+	m := &Modules{
+		cfg: cfg,
+		log: log.New("modules"),
+
+		grpcServer:                server,
+		kindRegistry:              kindRegistry,
+		backgroundServiceRegistry: backgroundServiceRegistry,
+	}
+	if err := roleRegistry.RegisterFixedRoles(context.Background()); err != nil {
+		panic(err)
+	}
+	return m
+}
+
+func (m *Modules) Init() error {
+	mm := modules.NewManager(m.log)
+	mm.RegisterModule(GRPCServer, m.initGRPCServer, modules.UserInvisibleModule)
```

Comment:
Nit. It would be nice if each service could register with the modules manager and provide its dependencies so that you have to re-implement all this code in Enterprise for example. Could create a chicken and egg problem perhaps, but if all of our services only instantiate, e.g. in wire Provide-function, and does nothing else (which is not true today, e.g. plugins registering during instantiation) it could allow service to take the module manager as dependency and register them self specifying dependencies to other services.